### DISCUSSION TITLE
Adding Cgreen into a CMake project as source from a git submodule (no system-wide install)

### DISCUSSION THREAD

**Discussion by DaStoned** [cgreen-devs/cgreen#289] [Q&A]
Hi! I'm considering Cgreen as the unit testing framework for a few C projects, but I'd very much prefer to not install it as a system-wide library. In this case I'd need to install and maintain project-specific versions of it in every developer machine, CI build agent, etc. Since this is a small and self-contained library, I'm looking to add it as a git submodule into every project which needs it; then just build and link it as a local library (static or shared, doesn't matter) when building the unit test executable. 

I'm struggling with the relevant top-level CMakeLists.txt (my CMake skills are quite weak). Does anyone have a practical example of how to pull Cgreen into CMakeLists.txt without a system-wide installation? E.g. maybe something like this (assuming the Cgreen source code is in a directory "unittest_source/cmake"), but actually working?

```
cmake_minimum_required(VERSION 3.15)
project(my_unittest)
add_executable(my_unittest ut_main.c)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/cgreen)
target_include_directories(my_unittest PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/cgreen/include)
target_link_directories(my_unittest PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/cgreen/src)
target_link_libraries(my_unittest cgreen)
```

This doesn't actually build Cgreen when building the unit test executable. If I go and manually trigger a build of Cgreen, the main executable still fails to link `-lcgreen`.

*Update* Arrived at this as a working solution:

```
cmake_minimum_required(VERSION 3.15)
project(my_unittest)
add_executable(my_unittest ut_main.c)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/cgreen)
target_include_directories(my_unittest PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/cgreen/include)
target_link_libraries(my_unittest cgreen_shared)
```

**Comment by thoni56** [comment]
I have no experience with hooking in a sub-CMake project. My experience with CMake is limited, not all good...

As this probably requires CMake specific knowledge, my suggestion would be to get some general advice on the matter from a CMake forum, or even StackOverflow, first.

**Reply by DaStoned** [reply]
Sure, that's solid advice. I can also go and improve my CMake skills. I was just hoping someone in the Cgreen community would have done this and has the answer ready :)

**Reply by thoni56** [reply]
Yes, sure. As a maintainer, I wanted it to be clear that _I_ had no experience with that.

Perhaps @snewton, the contributor of the CMake  `find-package` documentation in the README, has.

**Reply by DaStoned** [reply]
While reading through `cgreen/src/CMakeLists.txt` I understood that the source of my problem is in this command:
```target_link_libraries(my_unittest cgreen)```

Apparently Cgreen's build script doesn't create a library target named `cgreen`. Instead there are `cgreen_shared` and `cgreen_static` (thank you CMake for silently failing on me). Updated to this, it now compiles and links successfully:
```target_link_libraries(my_unittest cgreen_shared)```
